File: src/main/kotlin/leetcode/p0347_top_k_frequent_elements/solution.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
#include <math.h>

using namespace std;

class Solution {
   public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> map;

        for (auto i : nums) {
            map[i]++;
        }

        vector<vector<int>> bucket(nums.size() + 1);

        for (auto pair : map) {
            bucket[pair.second].push_back(pair.first);
        }

        vector<int> result;
        for (int i = nums.size(); i >= 0; i--) {
            if (!bucket[i].empty()) {
                int c = min(k, (int)bucket[i].size());
                result.insert(result.end(), bucket[i].begin(), bucket[i].begin() + c);
                k-=c;
                if (k == 0)
                    break;
            }
        }
        
        return result;
    }
};

int main() {
    Solution* s = new Solution();

    int k;
    cin >> k;

    int count;
    cin >> count;

    vector<int> vec;

    for (size_t i = 0; i < count; i++) {
        int l;
        cin >> l;
        vec.push_back(l);
    }

    auto res = s->topKFrequent(vec, k);

    for (size_t i = 0; i < res.size(); i++) {
        cout << res[i] << "\n";
    }

    return 0;
}

File: src/main/kotlin/leetcode/p0347_top_k_frequent_elements/solution.kt
package leetcode.p0347_top_k_frequent_elements

import kotlin.math.min

class Solution {
    fun topKFrequent(nums: IntArray, k: Int): IntArray {
        if (nums.size == 1) return nums

        val bucket = Array(nums.size + 1){
            ArrayList<Int>(min(10, nums.size/2))
        }
        HashMap<Int, Int>().let { map ->
            nums.forEach { map[it] = (map[it] ?: 0) + 1 }

            map.keys.forEach {
                val value = map[it]!!
                bucket[value].add(it)
            }
        }

        val result = IntArray(k)
        var index = 0
        for (i in nums.size downTo 0){
            if (bucket[i].size != 0)
                bucket[i].forEach {
                    result[index++] = it
                    if (index == k)
                        return result
                }
        }
        return result
    }
}